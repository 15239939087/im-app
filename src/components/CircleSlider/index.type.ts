export type CircularSliderProps = {
  /** *************************** Base ************************************** */
  /** 外圆的半径 */
  outerRadius?: number;
  /** 内圆的半径 */
  innerRadius?: number;
  /** 刻度值的宽度 */
  scaleWidth?: number;
  /** 滑道的宽度 */
  trackWidth?: number;
  /** 刻度值的长度 */
  scaleLength?: number;
  /** 手势响应的向内最大宽度 */
  responseWidth?: number;
  /** 激活轨道的宽度 */
  sliderWidth?: number;
  /** 激活轨道距离最外层的距离 默认30 */
  activeSliderDistance?: number;
  /** 当前值 */
  value?: number;
  /** 次要字体 */
  subFontValue?: string;
  /** 最小值 */
  min?: number;
  /** 最大值 */
  max?: number;

  /** *************************** action ************************************** */
  /** 离开响应 */
  onRelease?: (value: number) => any;
  /** 移动响应 */
  onMove?: (value: number) => any;

  /** *************************** style ************************************** */
  /** 背景颜色  */
  wrapperColor?: string;
  /** 滑轨颜色  */
  trackColor?: string;
  /** 滑条颜色  */
  sliderColor?: string;
  /** 刻度值颜色  */
  scaleColor?: string;
  /** 激活状态的刻度值颜色  */
  activeScaleColor?: string;
  /** 主文字颜色  */
  fontColor?: string;
  /** 主文字大小 */
  fontSize?: number;
  /** 次文字颜色  */
  subFontColor?: string;
  /** 次文字大小 */
  subFontSize?: number;
  /** 是否禁用 */
  disabled?: boolean;
};
