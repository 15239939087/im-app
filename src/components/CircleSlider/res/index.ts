const pointer = require('./pointer.png');

export { pointer };
