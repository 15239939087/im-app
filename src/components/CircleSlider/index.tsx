/* eslint-disable react/no-array-index-key */
import React, { useRef, useState, useMemo, useEffect } from 'react';
import {
  PanResponder,
  View,
  GestureResponderEvent,
  StyleSheet,
  Text as RNText,
  Image,
} from 'react-native';
import Svg, { Path, LinearGradient, Defs, Stop } from 'react-native-svg';
import extractBrush from 'react-native-svg/lib/module/lib/extract/extractBrush';
import { CircularSliderProps } from './index.type';
import {
  calcIsRange,
  getActiveSliderPath,
  getCurrentAngle,
  getMarkerScalePath,
  getPointerPosition,
  getRealValue,
  getSliderPath,
  getTargetAngle,
} from './utils';
import { pointer } from './res';

const CircularSlider: React.FC<CircularSliderProps> = props => {
  const {
    outerRadius = 145,
    wrapperColor = '#181c25',
    sliderColor = '#FF4545',
    trackColor = '#2D3749',
    scaleColor = '#2D3749',
    activeScaleColor = '#FF4545',
    value = 0,
    min = 0,
    max = 100,
    trackWidth = 5, // 滑道宽度
    sliderWidth = 15, // 轨道宽度
    scaleLength = 10,
    scaleWidth = 2,
    responseWidth = 20,
    activeSliderDistance = 30,
    fontColor = '#FF4545',
    subFontColor = 'white',
    fontSize = 80,
    subFontSize = 24,
    subFontValue,
    onMove,
    onRelease,
    disabled = false,
  } = props;

  let moving = false; // 是否可以移动
  let currentAngle = 0; // 当前角度
  let prevLocationX = 0; // 上一个位置的x坐标
  let prevLocationY = 0; // 上一个位置的y坐标
  let prevAngle = 0; // 上一个角度
  let overMaxRange = false; // 是否超出最大角度
  let overMinRange = false; // 是否超出最小角度

  let animateId = 0;
  let isAnimating = false;
  const activeRef: any = useRef();
  const scaleRefs: any = useRef([]);

  const terminalAngle = -350; // 刻度线显示的最大角度

  const trackRadius = outerRadius - activeSliderDistance; // 内滑道圆的半径

  const sliderRadius: number = outerRadius - activeSliderDistance - sliderWidth / 2; // 内滑条圆的半径

  const scaleRadius = outerRadius - activeSliderDistance / 2; // 刻度中心所在圆的半径

  const outerScaleRadius: number = scaleRadius + scaleLength / 2; // 刻度外圈圆的半径
  const innerScaleRadius: number = scaleRadius - scaleLength / 2; // 刻度内圈圆的半径

  const innerRadius = outerRadius - activeSliderDistance - sliderWidth / 2;

  const pointerRadius: number = innerRadius - 12; // 内圈指针的圆的半径

  const markerScaleCount = max > 40 ? max / 2 : max;
  const markerScales: any[] = new Array(markerScaleCount + 1).fill(
    (item: any, index: number) => index
  ); // 显示刻度的数组

  const [angle, setAngle] = useState(0);
  const [data, setData] = useState(getRealValue(min, max, value));

  const [pointerX, setPointX] = useState(0);
  const [pointerY, setPointY] = useState(0);
  const [pointerTransform, setPointerTransform] = useState(0);

  // 开始更新路径
  const startUpdatingAnimation = () => {
    if (!isAnimating) {
      currentAngle = 0;
      isAnimating = true;
      playUpdatingAnimation();
    }
  };

  const updateActiveSliderPath = () => {
    // 更新路径
    activeRef.current?.setNativeProps({
      d: getActiveSliderPath(
        currentAngle,
        sliderRadius + +sliderWidth / 2,
        terminalAngle,
        outerRadius
      ),
    });

    // 更新刻度
    markerScales.forEach((item, index) => {
      const { hasSlidered } = getPathRelative(currentAngle, index);
      scaleRefs?.current?.[index]?.setNativeProps({
        stroke: hasSlidered
          ? disabled
            ? extractBrush(scaleColor)
            : extractBrush(activeScaleColor)
          : extractBrush(scaleColor),
      });
    });
  };

  const updatingActiveSliderPath = () => {
    const targetAngle = getTargetAngle(value, min, max, terminalAngle);
    updateActiveSliderPath();
    if (currentAngle <= targetAngle) {
      currentAngle = targetAngle;
      stopAnimation();
    }
  };

  const playUpdatingAnimation = () => {
    const targetAngle = getTargetAngle(value, min, max, terminalAngle);
    const step = targetAngle / (500 / 20);
    updatingActiveSliderPath();
    const stepTerminalAngle = currentAngle + step;

    if (stepTerminalAngle <= targetAngle) {
      currentAngle = targetAngle;
    } else {
      currentAngle = stepTerminalAngle;
    }

    if (isAnimating && currentAngle >= terminalAngle) {
      animateId = requestAnimationFrame(playUpdatingAnimation);
    } else {
      stopAnimation();
    }
  };

  useEffect(() => {
    setData(value);
    setPointerPosition();
  }, [value]);

  useEffect(() => {
    startUpdatingAnimation();
  }, []);

  // 刻度值处理
  const getPathRelative = (_angle: number, index: number) => {
    const step = terminalAngle / markerScaleCount;
    const lineAngle = index * step;

    // 角度均为负数进行比较
    const hasSlidered = _angle <= lineAngle;
    const isLineAngle = Math.abs(lineAngle - _angle) <= 0.0001;

    return { hasSlidered, lineAngle, isLineAngle };
  };

  const stopAnimation = () => {
    if (isAnimating) {
      cancelAnimationFrame(animateId);
    }
    isAnimating = false;
  };

  const setPointerPosition = () => {
    const { x, y, transform } = getPointerPosition(
      value,
      min,
      max,
      terminalAngle,
      pointerRadius,
      outerRadius
    );
    setPointX(x);
    setPointY(y);
    setPointerTransform(transform);
  };
  const stopMove = (realValue: number, isMax = false, isEnd: boolean, callback) => {
    stopAnimation();
    setAngle(isMax ? terminalAngle : currentAngle);
    callback(isMax ? max : realValue);
  };

  // 越界处理
  const setIsOverRange = (locationX: number, locationY: number) => {
    const _currentAngle = getCurrentAngle(locationX, locationY, outerRadius);
    if (prevLocationX && prevLocationY) {
      const dx = locationX - prevLocationX;
      const dy = locationY - prevLocationY;

      // 顺时针旋转越界
      if (
        (prevAngle !== undefined &&
          prevAngle <= -310 &&
          prevAngle >= terminalAngle &&
          dx >= 0 &&
          _currentAngle - prevAngle > 300) ||
        overMaxRange
      ) {
        if (!overMaxRange) {
          overMaxRange = true;
          currentAngle = -350;
          stopMove(max, true, false, onMove);
        }
      }
      // 解除顺时针旋转越界
      if (
        _currentAngle >= terminalAngle &&
        _currentAngle <= -310 &&
        dx <= 0 &&
        dy >= 0 &&
        overMaxRange
      ) {
        overMaxRange = false;
      }

      // 逆时针越界
      if (
        _currentAngle >= terminalAngle &&
        _currentAngle <= -310 &&
        dx <= 0 &&
        dy >= 0 &&
        overMaxRange
      ) {
        overMaxRange = false;
      }

      // 逆时针旋转越界
      if (
        (prevAngle !== undefined &&
          prevAngle <= 0 &&
          prevAngle >= -60 &&
          dx <= 0 &&
          prevAngle - _currentAngle > 300) ||
        overMinRange
      ) {
        if (!overMinRange) {
          overMinRange = true;
          currentAngle = 0;
          stopMove(min, false, false, onMove);
        }
      }

      // 解除逆时针越界
      if (_currentAngle <= 0 && _currentAngle >= -60 && dx >= 0 && dy >= 0 && overMinRange) {
        overMinRange = false;
      }
    }
  };

  // 移动处理
  const handleMove = (e: GestureResponderEvent, isEnd: boolean, func: any) => {
    const { locationX, locationY } = e.nativeEvent;
    if (!isEnd || moving) {
      setIsOverRange(locationX, locationY);
    } else {
      overMinRange = false;
      overMaxRange = false;
    }

    prevLocationX = locationX;
    prevLocationY = locationY;
    if (overMinRange || overMaxRange) {
      stopAnimation();
      return;
    }
    const moveAngle = Math.max(getCurrentAngle(locationX, locationY, outerRadius), terminalAngle);

    // 根据角度计算出的值
    const currentValue = min + (moveAngle * (max - min)) / terminalAngle;
    // 将该值进行四舍五入，获取到真正想要的整数值
    const realValue = Math.round(currentValue);
    currentAngle = Math.max((terminalAngle * (realValue - min)) / (max - min), terminalAngle);
    updateActiveSliderPath();
    prevAngle = currentAngle;
    if (currentAngle >= terminalAngle) {
      stopMove(realValue, false, isEnd, func);
    } else {
      stopMove(realValue, true, isEnd, func);
    }
  };

  const handleStartPanResponder = (e: GestureResponderEvent) => {
    const { locationX, locationY } = e.nativeEvent;
    const isRange = calcIsRange(locationX, locationY, trackRadius, outerRadius, responseWidth);

    if (isRange) {
      e.persist();
      return true;
    }
    return false;
  };

  const handleSetPanResponder = (e: GestureResponderEvent) => {
    const { locationX, locationY } = e.nativeEvent;
    const isRange = calcIsRange(locationX, locationY, trackRadius, outerRadius, responseWidth);
    const _currentAngle = getCurrentAngle(locationX, locationY, outerRadius);

    if (isRange) {
      e.persist();
      return true;
    }
    if (_currentAngle > 0 && _currentAngle < terminalAngle) {
      return false;
    }
    return false;
  };

  const handleTerminate = () => {
    moving = false;
  };

  // 手势移动
  const _onMove = (e: GestureResponderEvent) => {
    moving = true;
    handleMove(e, false, onMove);
  };

  // 手势点击
  const _onRelease = (e: GestureResponderEvent) => {
    handleMove(e, true, onRelease);
    moving = false;
  };

  // 手势系统
  const panResponder = useMemo(
    () =>
      PanResponder.create({
        onStartShouldSetPanResponder: handleStartPanResponder,
        onMoveShouldSetPanResponder: handleSetPanResponder,
        onPanResponderTerminationRequest: () => !moving,

        onPanResponderMove: _onMove,
        onPanResponderRelease: _onRelease,

        // 当前有其他的东西成为响应器并且没有释放它。
        onPanResponderReject: handleTerminate,
        onPanResponderTerminate: handleTerminate,
      }),
    [moving]
  );

  const markerScalesJSX = () => {
    return markerScales.map((scale, i) => {
      const { hasSlidered, lineAngle } = getPathRelative(angle, i);
      return (
        <Path
          key={i}
          d={getMarkerScalePath(lineAngle, innerScaleRadius, outerScaleRadius, outerRadius)}
          strokeWidth={scaleWidth}
          fill="transparent"
          stroke={hasSlidered ? activeScaleColor : scaleColor}
          ref={node => {
            scaleRefs.current[i] = node;
          }}
        />
      );
    });
  };

  return (
    <View
      pointerEvents={disabled ? 'none' : 'box-only'}
      style={{
        width: outerRadius * 2,
        height: outerRadius * 2,
        justifyContent: 'center',
        alignItems: 'center',
      }}
      {...panResponder.panHandlers}
    >
      <View
        style={{
          width: outerRadius * 2,
          height: outerRadius * 2,
          backgroundColor: wrapperColor,
          position: 'absolute',
          borderRadius: outerRadius * 2,
        }}
      />

      <Svg width={outerRadius * 2} height={outerRadius * 2}>
        <Defs>
          <LinearGradient id="grad_2" x1="0" y1="0" x2="1" y2="1">
            <Stop offset="0" stopColor={trackColor} stopOpacity="1" />
            <Stop offset="1" stopColor={trackColor} stopOpacity="1" />
          </LinearGradient>
        </Defs>
        <Path
          d={getSliderPath(0, trackRadius, outerRadius, terminalAngle)}
          strokeWidth={trackWidth}
          stroke="url(#grad_2)"
          fill="transparent"
        />
        {!disabled && (
          <Path
            style={{ zIndex: 999 }}
            d={getActiveSliderPath(
              angle,
              sliderRadius + sliderWidth / 2,
              terminalAngle,
              outerRadius
            )}
            strokeWidth={sliderWidth}
            fill="none"
            stroke={sliderColor}
            ref={activeRef}
          />
        )}

        {/** 刻度值 */}
        {markerScalesJSX()}
      </Svg>

      <View
        style={[
          styles.innerWrapper,
          {
            width: innerRadius * 2,
            height: innerRadius * 2,
            borderRadius: innerRadius * 2,
            top: outerRadius - innerRadius,
            left: outerRadius - innerRadius,
          },
        ]}
        pointerEvents="none"
      >
        <View style={[styles.textWrapper, { opacity: disabled ? 0.25 : 1 }]}>
          <RNText style={[styles.mainValueText, { color: fontColor, fontSize }]}>{data}</RNText>
          <RNText
            style={{
              color: subFontColor,
              fontSize: subFontSize,
              fontWeight: '400',
              marginTop: -16,
              fontFamily: 'Helvetica',
            }}
          >
            {subFontValue}
          </RNText>
        </View>
      </View>
      {/* 指针 */}
      {!disabled && (
        <Image
          source={pointer}
          style={[
            styles.pointerStyle,
            {
              left: pointerX,
              top: pointerY,
              transform: [{ rotate: `${pointerTransform}deg` }],
              zIndex: 3,
            },
          ]}
        />
      )}
    </View>
  );
};
const styles = StyleSheet.create({
  innerWrapper: {
    alignItems: 'center',
    backgroundColor: '#2D3749',
    flex: 1,
    justifyContent: 'center',
    left: 0,
    position: 'absolute',
    top: 0,
    zIndex: 2,
  },
  mainValueText: {
    fontFamily: 'Helvetica',
    fontWeight: '600',
  },
  pointerStyle: {
    height: 12,
    position: 'absolute',
    width: 13,
  },

  textWrapper: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default CircularSlider;
