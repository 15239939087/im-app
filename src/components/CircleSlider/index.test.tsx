import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import CircleSlider from './index';

describe('<CircleSlider />', () => {
  it('render CircleSlider', () => {
    const wrapper = shallow(
      <CircleSlider
        outerRadius={145}
        wrapperColor="#23242e"
        fontSize={108}
        fontColor="white"
        subFontColor="white"
        subFontValue="current 18â„ƒ"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
