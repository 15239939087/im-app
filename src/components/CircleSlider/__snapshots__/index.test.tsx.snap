// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CircleSlider /> render CircleSlider 1`] = `
<View
  onLayout={[Function]}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  pointerEvents="box-only"
  style={
    Object {
      "alignItems": "center",
      "height": 290,
      "justifyContent": "center",
      "width": 290,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#23242e",
        "borderRadius": 290,
        "height": 290,
        "position": "absolute",
        "width": 290,
      }
    }
  />
  <Svg
    height={290}
    preserveAspectRatio="xMidYMid meet"
    width={290}
  >
    <Defs>
      <LinearGradient
        id="grad_2"
        x1="0"
        x2="1"
        y1="0"
        y2="1"
      >
        <Stop
          offset="0"
          stopColor="#2D3749"
          stopOpacity="1"
        />
        <Stop
          offset="1"
          stopColor="#2D3749"
          stopOpacity="1"
        />
      </LinearGradient>
    </Defs>
    <Path
      d="M145 30 A 115 115, 0, 0, 1, 260 145 A 115 115, 0, 1, 1, 125.0304595683029 31.747108403596087"
      fill="transparent"
      stroke="url(#grad_2)"
      strokeWidth={5}
    />
    <Path
      d="M145 30 A 115 115, 0, 0, 1, 145 30"
      fill="none"
      stroke="#FF4545"
      strokeWidth={15}
      style={
        Object {
          "zIndex": 999,
        }
      }
    />
    <Path
      d="M145 20 L145 10"
      fill="transparent"
      key="0"
      stroke="#FF4545"
      strokeWidth={2}
    />
    <Path
      d="M160.23366792564343 20.93173104483475 L161.45236135969492 11.00626952842154"
      fill="transparent"
      key="1"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M175.24023694995847 23.713034215500443 L177.65945590595516 14.010076952740462"
      fill="transparent"
      key="2"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M189.79599369316253 28.30244668784978 L193.37967318861553 18.96664242287777"
      fill="transparent"
      key="3"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M203.68394534823636 34.631550892634124 L208.37866097609526 25.802074964044863"
      fill="transparent"
      key="4"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M216.69705454388077 42.60599446387603 L222.4328189073912 34.414474020986106"
      fill="transparent"
      key="5"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M228.64132579485727 52.106896815325726 L235.33263185844586 44.67544856055177"
      fill="transparent"
      key="6"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M239.3386975278465 62.99262137618659 L246.88579333007422 56.43203108628151"
      fill="transparent"
      key="7"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M248.6296965693802 75.10088706615664 L256.9200722949306 69.50895803144918"
      fill="transparent"
      key="8"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M256.37581552354595 88.25118753255664 L265.2858807654296 83.71128253516117"
      fill="transparent"
      key="9"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M262.46157759823853 102.24748208429139 L271.8585038060976 98.8272806510347"
      fill="transparent"
      key="10"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M266.7962580981544 116.88111820701688 L276.53995874600673 114.63160766357824"
      fill="transparent"
      key="11"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M269.31523692103417 131.93394209154332 L279.26045587471685 130.8886574588668"
      fill="transparent"
      key="12"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M269.9809618945489 147.18155080466045 L279.9794388461128 147.35607486903328"
      fill="transparent"
      key="13"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M268.7835085926963 162.39663762000816 L278.686189280112 163.7883686296088"
      fill="transparent"
      key="14"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M265.74072828613356 177.35238063781512 L275.3999865490242 179.94057108884033"
      fill="transparent"
      key="15"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M260.89798182084843 191.82582417698902 L270.1698203665163 195.57189011114812"
      fill="transparent"
      key="16"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M254.32746339242448 205.60120253079214 L263.07366046381844 210.4492987332555"
      fill="transparent"
      key="17"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M246.12712429686843 218.47315653655915 L254.2172942406179 224.35100905948386"
      fill="transparent"
      key="18"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M236.41921270239632 230.24979500781228 L243.73274971858802 237.06977860843728"
      fill="transparent"
      key="19"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M225.34845121081742 240.75555538987226 L231.77632730768283 248.41599982106203"
      fill="transparent"
      key="20"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M213.07987937687835 249.83382099317802 L218.52626972702862 258.2205266726323"
      fill="transparent"
      key="21"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M199.79639334863467 257.34925578739586 L204.18010481652544 266.33719625038754"
      fill="transparent"
      key="22"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M185.69601930714464 263.1898219499146 L188.9517008517162 272.64500770590774"
      fill="transparent"
      key="23"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M170.98896135221992 267.2684500917257 L173.0680782603975 277.0499260990638"
      fill="transparent"
      key="24"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M155.89446784345733 269.5243372614682 L156.7660252709339 279.48628424238564"
      fill="transparent"
      key="25"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M140.63756291218738 269.923853377387 L140.28856794516238 279.9177616475779"
      fill="transparent"
      key="26"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M125.44569186997116 268.4610425743922 L123.88134721956885 278.3379259803436"
      fill="transparent"
      key="27"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M110.54533052287513 265.1577119922899 L107.78895696470514 274.77032895167304"
      fill="transparent"
      key="28"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M96.15860893884081 260.06310668155504 L92.25129765394807 269.26815521607944"
      fill="transparent"
      key="29"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M82.49999999999999 253.25317547305482 L77.49999999999999 261.9134295108992"
      fill="transparent"
      key="30"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M69.77312210599399 244.8294387559116 L63.75497187447351 252.81579385638457"
      fill="transparent"
      key="31"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M58.167703692625324 234.91747504233138 L51.221119988035355 242.11087304571788"
      fill="transparent"
      key="32"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M47.85675481787861 223.66504888122964 L40.0852952033089 229.958252791728"
      fill="transparent"
      key="33"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M38.99398798044676 211.23990802915063 L30.513507018882493 216.5391006714827"
      fill="transparent"
      key="34"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M31.71152662041878 197.82728271758748 L22.64844875005228 202.05346533499448"
      fill="transparent"
      key="35"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M26.11793546310581 183.62712429686843 L16.607370300154287 186.71729424061792"
      fill="transparent"
      key="36"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M22.296602069042024 168.85112442206818 L12.480330234565372 170.75921437583364"
      fill="transparent"
      key="37"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M20.30449371752198 153.7195592180157 L10.328853214923726 154.41712395545696"
      fill="transparent"
      key="38"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M20.171308155678275 138.458005469632 L10.18501280813254 137.93464590720257"
      fill="transparent"
      key="39"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M21.89903087347399 123.29397779163375 L12.050953343351892 121.55749601496446"
      fill="transparent"
      key="40"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M25.461905504620574 108.45353690965791 L15.898857944990226 105.52981986243054"
      fill="transparent"
      key="41"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M30.8068177946749 94.15791961552493 L21.6713632182489 90.09055318476692"
      fill="transparent"
      key="42"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M37.854087412235955 80.62024063624322 L29.282414405214837 75.46985988714269"
      fill="transparent"
      key="43"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M46.49865579915978 68.04231558429268 L38.61854826309256 61.885700831036104"
      fill="transparent"
      key="44"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M56.61165235168154 56.611652351681585 L49.54058453981607 49.54058453981611"
      fill="transparent"
      key="45"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M68.04231558429264 46.498655799159806 L61.88570083103606 38.61854826309259"
      fill="transparent"
      key="46"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M80.6202406362432 37.854087412235984 L75.46985988714265 29.282414405214865"
      fill="transparent"
      key="47"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M94.15791961552497 30.806817794674885 L90.09055318476697 21.67136321824887"
      fill="transparent"
      key="48"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M108.45353690965797 25.461905504620546 L105.5298198624306 15.898857944990198"
      fill="transparent"
      key="49"
      stroke="#2D3749"
      strokeWidth={2}
    />
    <Path
      d="M123.2939777916336 21.899030873474018 L121.55749601496427 12.050953343351921"
      fill="transparent"
      key="50"
      stroke="#2D3749"
      strokeWidth={2}
    />
  </Svg>
  <View
    pointerEvents="none"
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2D3749",
          "flex": 1,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "top": 0,
          "zIndex": 2,
        },
        Object {
          "borderRadius": 215,
          "height": 215,
          "left": 37.5,
          "top": 37.5,
          "width": 215,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Helvetica",
              "fontWeight": "600",
            },
            Object {
              "color": "white",
              "fontSize": 108,
            },
          ]
        }
      >
        0
      </Text>
      <Text
        style={
          Object {
            "color": "white",
            "fontFamily": "Helvetica",
            "fontSize": 24,
            "fontWeight": "400",
            "marginTop": -16,
          }
        }
      >
        current 18℃
      </Text>
    </View>
  </View>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 12,
          "position": "absolute",
          "width": 13,
        },
        Object {
          "left": 0,
          "top": 0,
          "transform": Array [
            Object {
              "rotate": "0deg",
            },
          ],
          "zIndex": 3,
        },
      ]
    }
  />
</View>
`;
